bert_trainer.py

```diff
@@ -0,0 +1,259 @@
+
+                if mode == tf.estimator.ModeKeys.TRAIN:
+                    return tf.estimator.EstimatorSpec(mode=mode,
+                            loss=loss,
+                            train_op=train_op)
+                else:
+                    return tf.estimator.EstimatorSpec(mode=mode,
+                            loss=loss,
+                            eval_metric_ops=eval_metrics)
+            else:
+                (predicted_labels, log_probs) = self._create_model(
+                        is_predicting, input_ids, input_mask, segment_ids, label_ids, num_labels)
+
+                predictions = {
+                        'probabilities': log_probs,
+                        'labels': predicted_labels
+                        }
+                return tf.estimator.EstimatorSpec(mode, predictions=predictions)
+
+        # Return the actual model function in the closure
+        return model_fn
+
+
+    def _create_estimator(self):
+        # Compute train and warmup steps from batch size
+        self.num_train_steps = int(
+            len(self.train_features) / self.batch_size * self.num_train_epochs)
+        self.num_warmup_steps = int(
+            self.num_train_steps * self.warmup_proportion)
+
+        # Specify output directory and number of checkpoint steps to save
+        run_config = tf.estimator.RunConfig(
+                model_dir=self.output_dir,
+                save_summary_steps=self.save_summary_every,
+                save_checkpoints_steps=self.save_checkpoints_every)
+
+        model_fn = self._model_fn_builder(
+                num_labels=len(self.label_list),
+                learning_rate=self.learning_rate,
+                num_train_steps=self.num_train_steps,
+                num_warmup_steps=self.num_warmup_steps)
+
+        self.estimator = tf.estimator.Estimator(
+                model_fn=model_fn,
+                config=run_config,
+                warm_start_from=self.checkpoint_dir,
+                params={"batch_size": self.batch_size})
+
+    # Create an input function for training. drop_remainder = True for using TPUs.
+    def train_model(self):
+        train_input_fn = bert.run_classifier.input_fn_builder(
+                features=self.train_features,
+                seq_length=self.max_seq_length,
+                is_training=True,
+                drop_remainder=False)
+
+        print(f'Beginning Training!')
+        current_time = datetime.now()
+        self.estimator.train(input_fn=train_input_fn, max_steps=self.num_train_steps)
+        print("Training took time ", datetime.now() - current_time)
+
+    def test_model(self):
+        test_input_fn = run_classifier.input_fn_builder(
+                features=self.test_features,
+                seq_length=self.max_seq_length,
+                is_training=False,
+                drop_remainder=False)
+
+        return self.estimator.evaluate(input_fn=test_input_fn, steps=None)
+
+    def predict(self, inputs):
```

Comment:
add some comments about the inputs

http://google.github.io/styleguide/pyguide.html#383-functions-and-methods